database/sqlserver/sqlserver.go

```diff
@@ -116,16 +119,43 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
 	return ss, nil
 }
 
-// Open a connection to the database
+// Open a connection to the database.
 func (ss *SQLServer) Open(url string) (database.Driver, error) {
 	purl, err := nurl.Parse(url)
 	if err != nil {
 		return nil, err
 	}
 
-	db, err := sql.Open("sqlserver", migrate.FilterCustomQuery(purl).String())
-	if err != nil {
-		return nil, err
+	useMsiParam := purl.Query().Get("useMsi")
+	useMsi := false
+	if len(useMsiParam) > 0 {
+		useMsi, err = strconv.ParseBool(useMsiParam)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var db *sql.DB
+	if useMsi {
+		tokenProvider, err := getMSITokenProvider(fmt.Sprintf("%s%s", "https://", strings.Join(strings.Split(purl.Hostname(), ".")[1:], ".")))
```

Comment:
Could this line be refactored into a helper in order to make its purpose more clear and maybe make it testable? I'm a little worried about some of these string operations being brittle with unexpected input.

Reply:
yeah, I agree. The main purpose here is to get the resource uri which could change across clouds. I'll see what I can do. Currently, if the uri is incorrect, I believe it would result in a 4xx (invalid resource uri for tenant or something similar). Will validate that behavior and update

Reply:
Sounds good, thank you! The cloud change use case is not something I had picked up from reading the code but that's super important.